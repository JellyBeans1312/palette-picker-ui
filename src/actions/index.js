export const addProject = id => ({
  type: 'ADD_PROJECT',
  id
});

export const addAllProjects = allProjects => ({
  type: 'ADD_ALL_PROJECTS',
  allProjects
});

export const saveColors = colors => ({
  type: 'SAVE_COLORS',
  colors
});