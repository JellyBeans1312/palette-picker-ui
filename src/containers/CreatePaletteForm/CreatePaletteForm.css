.palette-form {
  background-color: red;
  height: 50%;
  width: 100%;
}