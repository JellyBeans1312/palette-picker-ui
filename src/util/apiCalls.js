export const createProject = async (newProject) => {
  const options = {
    method: "POST",
    body: JSON.stringify(newProject),
    headers: {
      "Content-Type": "application/json"
    }
  }
  const response = await fetch('https://palette-picker-be-eo-am.herokuapp.com/api/v1/projects', options)
  if(!response.ok && response.status === 409) {
    throw Error('That project name already exists. Please choose another name.')
  } else if (!response.ok) {
    throw Error('There was an issue creating your project. Please try again.')
  }

  const data = await response.json();
  return data;
}

export const getAllProjects = async () => {
  const response = await fetch('https://palette-picker-be-eo-am.herokuapp.com/api/v1/projects');
  if(!response.ok) {
    throw Error('There was an error retrieving your projects. Please try again.')
  }
  const data = await response.json()
  return data;
}

export const getAllPalettes = async () => {
  const response = await fetch('https://palette-picker-be-eo-am.herokuapp.com/api/v1/palettes')
  if(!response.ok) {
    throw Error('There was an error retrieving your palettes. Please try again.')
  }
  const data = await response.json();
  return data;
}

export const createPalette = async (newPalette) => {
  console.log('NEW PALETTE', newPalette)
  const options = {
    method: "POST",
    body: JSON.stringify(newPalette),
    headers: {
      "Content-Type": "application/json"
    }
  }
  console.log('OPTIONS ARE HERE', options)
  const response = await fetch('https://palette-picker-be-eo-am.herokuapp.com/api/v1/palettes', options)
  console.log('RESPONSE IS HERE', response)
  if(!response.ok) {
    throw Error('There was an error creating your palette. Please try again.')
  }
  const data = await response.json();
  return data;
}

export const deletePalette = async (id) => {
  const options = {
    method: 'DELETE'
  }

  const response = await fetch(`https://palette-picker-be-eo-am.herokuapp.com/api/v1/palettes/${id}`, options)
  if(!response.ok) {
    throw Error('There was an error deleting your palette. Please try again.')
  }

  return response.status;
}

export const deleteProject = async (id) => {
  const options = {
    method: 'DELETE'
  }
  const response = await fetch(`https://palette-picker-be-eo-am.herokuapp.com/api/v1/projects/${id}`, options)
  if(!response.ok) {
    throw Error('There was an error deleting your project. Please try again.')
  }
  return response.status
}