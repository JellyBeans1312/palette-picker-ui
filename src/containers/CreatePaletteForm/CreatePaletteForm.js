import React, { Component } from 'react';
import { createPalette, getAllProjects, getAllPalettes } from '../../util/apiCalls';
import { addAllProjects, addAllPalettes } from '../../actions/index';
import { connect } from 'react-redux';
import './CreatePaletteForm.css';

export class CreatePaletteForm extends Component {
  constructor() {
    super();
    this.state = {
      paletteName: ''
    }
  }

  handleChange = (e) => {
    this.setState({[e.target.name]: e.target.value})
  }

  savePalette = async (e) => {
    e.preventDefault();
    const { project, colors, addAllProjects, addAllPalettes } = this.props;
    const newPalette = {
      project_id: project.id,
      palette_name: this.state.paletteName,
      color_one: colors[0].color,
      color_two: colors[1].color,
      color_three: colors[2].color,
      color_four: colors[3].color,
      color_five: colors[4].color
    }
    
    createPalette(newPalette);
    const allProjects = await getAllProjects();
    addAllProjects(allProjects);
    const allPalettes = await getAllPalettes();
    addAllPalettes(allPalettes);
    this.clearInput();
  }
  
  clearInput = () => {
    this.setState({paletteName: ''});
  }

  render() {
    return (
      <form>
        <input
          type = 'text'
          value = {this.state.paletteName}
          onChange = {this.handleChange}
          placeholder = 'Enter a palette name!'
          name = 'paletteName'
        >
        </input>
        {this.props.project && this.state.paletteName &&
        <button
          onClick={(e) => this.savePalette(e)}
        >
        Save palette to current project.
        </button>
        }
      </form>
    )
  }
}

export const mapDispatchToProps = dispatch => ({
  addAllProjects: allProjects => dispatch(addAllProjects(allProjects)),
  addAllPalettes: allPalettes => dispatch(addAllPalettes(allPalettes))
});

export const mapStateToProps = state => ({
  colors: state.colors,
  project: state.project
});

export default connect(mapStateToProps, mapDispatchToProps)(CreatePaletteForm);