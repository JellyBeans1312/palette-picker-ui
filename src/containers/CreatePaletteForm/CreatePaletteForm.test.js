import React from 'react';
import { shallow } from 'enzyme';
import { CreatePaletteForm, mapDispatchToProps, mapStateToProps } from './CreatePaletteForm';
import { addAllProjects, addAllPalettes, addPalette, addProject } from '../../actions';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { createPalette, getAllProjects, getAllPalettes, searchSpecificPalette } from '../../util/apiCalls';

configure({ adapter: new Adapter() });

jest.mock('../../util/apiCalls')

describe('Create Palette Form', () => {
  let wrapper;
  const mockColors = [
    {
      color:"#fff",
      isLocked:false
    },
    {
      color:"#000",
      isLocked:false
    },
    {
      color:"#ececec",
      isLocked:false
    },
    {
      color:"#aeaeae",
      isLocked:false
    },
    {
      color:"#d8d646",
      isLocked:false
    }
]

  const mockPalette = {
    id:24,
    palette_name:"sdf",
    project_id:50,
    color_one:"#67cd51",
    color_two:"#15fe0d",
    color_three:"#507dbb",
    color_four:"#784994",
    color_five:"#2570dd",
    created_at:"2019-10-15T01:54:56.793Z",
    updated_at:"2019-10-15T01:54:56.793Z",
  }
  const mockProject = {
    id: 50,
    project_name:"asdf"
  }
  const mockAllProjects=[{
    id:50,
    project_name:"Aidan's fantastic project",
    created_at:"2019-10-15T02:16:24.923Z",
    updated_at:"2019-10-15T02:16:24.923Z"
  }]

  let mockEvent;

  beforeAll(() => {
    searchSpecificPalette.mockImplementation(() => mockPalette)
    createPalette.mockImplementation(() => mockPalette.id)
    getAllPalettes.mockImplementation(() => mockPalette)
    getAllProjects.mockImplementation(() => mockAllProjects)

  })

  const mockAddProject = jest.fn()
  const mockAddAllPalettes = jest.fn()
  const mockAddAllProjects = jest.fn()
  beforeEach(() => {
    mockEvent = { preventDefault: jest.fn()}
    wrapper = shallow(
      <CreatePaletteForm
        colors={mockColors}
        project={mockProject}
        allProjects={mockAllProjects}
        addAllPalettes={mockAddAllPalettes}
        addAllProjects={mockAddAllProjects}
        addPalette={jest.fn()}
        addProject={mockAddProject}
      />
    )
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  });

  it('should call handleChange on change of palette name input and update state', () => {
    const mockEvent = {
      target: { value: 'palette', name:'paletteName' }
    }
    wrapper.find('.form-input').at(0).simulate('change', mockEvent);
    expect(wrapper.state('paletteName')).toEqual('palette')
  });

  it('should call handleChange on change of palette name search and update state', () => {
    const mockEvent = {
      target: { value: 'palette', name:'searchPaletteName' }
    }
    wrapper.find('.form-input').at(1).simulate('change', mockEvent);
    expect(wrapper.state('searchPaletteName')).toEqual('palette')
  });

  it('should call addAll projects, addAllPalettes', async () => {
    await wrapper.instance().savePalette(mockEvent);
    expect(mockAddAllPalettes).toHaveBeenCalled();
    expect(mockAddAllProjects).toHaveBeenCalled();
  })

  it('should call addProject when search palette is invoked', async () => {
    await wrapper.instance().searchPalette(mockEvent);

    expect(mockAddProject).toHaveBeenCalled()
  });

  describe("mapStateToProps", () => {
    it("should return a project with id", () => {
      const mockState = {
        project: mockProject
      };
      const expected = {
        project: mockProject
      };
      const mappedProps = mapStateToProps(mockState);

      expect(mappedProps).toEqual(expected);
    });

    it("should return an array with colors", () => {
      const mockState = {
        colors: mockColors
      };
      const expected = {
        colors: mockColors
      };

      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });

    it("should return an array with projects", () => {
      const mockState = {
        allProjects: mockAllProjects
      };
      const expected = {
        allProjects: mockAllProjects
      };

      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });
  });

  describe("mapDispatchToProps", () => {
    it("calls dispatch with an addAllProjects action", () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = addAllProjects(mockAllProjects);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.addAllProjects(mockAllProjects);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
    it("calls dispatch with an addAllPalettes action", () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = addAllPalettes(mockColors);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.addAllPalettes(mockColors);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
    it("calls dispatch with an addPalette action", () => {
      const mockPalette = {
        id:21,
        palette_name:"asdf",
        project_id:50,
        color_one:"#67cd51",
        color_two:"#15fe0d",
        color_three:"#507dbb",
        color_four:"#784994",
        color_five:"#2570dd",
        created_at:"2019-10-15T01:54:46.520Z",
        updated_at:"2019-10-15T01:54:46.520Z"
      }
      const mockDispatch = jest.fn();
      const actionToDispatch = addPalette(mockPalette);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.addPalette(mockPalette);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
      it("calls dispatch with an addProject action", () => {
        const mockDispatch = jest.fn();
        const actionToDispatch = addProject(mockProject);
        const mappedProps = mapDispatchToProps(mockDispatch);
        mappedProps.addProject(mockProject);
  
        expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
      })
    })
})