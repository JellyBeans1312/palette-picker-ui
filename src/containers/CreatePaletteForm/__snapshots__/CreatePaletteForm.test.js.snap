// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Palette Form should match snapshot 1`] = `
<form
  className="palette-form"
>
  <input
    className="form-input"
    name="paletteName"
    onChange={[Function]}
    placeholder="Enter a palette name!"
    type="text"
    value=""
  />
  <input
    className="form-input"
    name="searchPaletteName"
    onChange={[Function]}
    placeholder="Search a for a palette"
    type="text"
    value=""
  />
  <button
    className="palette-button"
    onClick={[Function]}
  >
    Search!
  </button>
</form>
`;
